File: leetcode/Edit_Distance.java
import java.util.*;

public class Solution {
    static int[][] d = new int[501][501];
    public int minDistance(String word1, String word2) {
        if (word1.length() == 0) return word2.length();
        if (word2.length() == 0) return word1.length();

        for (int i = 0; i <= word1.length(); ++i)
            for (int j = 0; j <= word2.length(); ++j)
                d[i][j] = Integer.MAX_VALUE;
        d[0][0] = 0;

        for (int i = 0; i <= word1.length(); ++i)
            for (int j = 0; j <= word2.length(); ++j) {
                if (i < word1.length() && j < word2.length()) {
                    if (word1.charAt(i) == word2.charAt(j)) {
                        if (d[i + 1][j + 1] > d[i][j]) d[i + 1][j + 1] = d[i][j];
                    } else {
                        if (d[i + 1][j + 1] > d[i][j] + 1) d[i + 1][j + 1] = d[i][j] + 1;
                    }
                }
                if (i < word1.length() && d[i + 1][j] > d[i][j] + 1) {
                    d[i + 1][j] = d[i][j] + 1;
                }
                if (j < word2.length() && d[i][j + 1] > d[i][j] + 1) {
                    d[i][j + 1] = d[i][j] + 1;
                }
            }

        return d[word1.length()][word2.length()];
    }
}

File: leetcode/Edit_Distance.cpp
typedef pair<string, int> PSI;

class Solution {
    map<PSI, int> cache;
    int measure(string s, int k, string t) {
        if (cache.find(make_pair(s, k)) != cache.end())
            return cache[make_pair(s, k)];
        if (k == t.length()) {
            int distance = s.length();
            cache[make_pair(s, k)] = distance;
            return distance;
        }
        if (s == "") {
            int distance = t.length() - k;
            cache[make_pair(s, k)] = distance;
            return distance;
        }
        if (s[0] == t[k]) {
            int distance = measure(s.substr(1), k + 1, t);
            cache[make_pair(s, k)] = distance;
            return distance;
        } else {
            int min_distance = 987654321;
            int distance = 1 + measure(s, k + 1, t);  // Insert
            if (min_distance > distance)
                min_distance = distance;
            distance = 1 + measure(s.substr(1), k + 1, t);  // Replace
            if (min_distance > distance)
                min_distance = distance;
            distance = 1 + measure(s.substr(1), k, t);  // Delete
            if (min_distance > distance)
                min_distance = distance;
            cache[make_pair(s, k)] = min_distance;
            return min_distance;
        }
    }
public:
    int minDistance(string word1, string word2) {
        cache.clear();
        return measure(word1, 0, word2);
    }
};
